### PR TITLE
Pin 1315 authorization-process GET /clients

### PR DESCRIPTION
**Pull request by nttdata-rtorsoli**

### CODE REVIEW

File: src/main/resources/interface-specification.yml [hunk 1]
```diff
@@ -100,41 +100,41 @@ paths:
...
+              format: uuid
+            default: []
+          explode: false
+        - in: query
+          name: offset
+          required: true
           schema:
             type: integer
             format: int32
-        - name: consumerId
```
Comment (galales) [inline src/main/resources/interface-specification.yml:117]: we should keep the existing filters for services that use them

File: src/main/resources/interface-specification.yml [hunk 2]
```diff
@@ -100,41 +100,41 @@ paths:
...
-          in: query
-          description: ID of Consumer that MUST be related to the Client
+            minimum: 0
+        - in: query
+          name: limit
           required: true
           schema:
-            type: string
-            format: uuid
-        - name: purposeId
```
Comment (galales) [inline src/main/resources/interface-specification.yml]: same, let's keep this filter

File: src/main/resources/interface-specification.yml [hunk 3]
```diff
@@ -100,41 +100,41 @@ paths:
...
-          in: query
-          description: ID of Purpose that MUST be related to the Client
-          schema:
-            type: string
-            format: uuid
-        - name: kind
```
Comment (galales) [inline src/main/resources/interface-specification.yml]: same, let's keep this filter

File: src/main/resources/interface-specification.yml [hunk 4]
```diff
@@ -836,22 +830,58 @@ components:
...
+      type: object
+      properties:
+        id:
+          type: string
+          format: uuid
+        name:
+          type: string
+        description:
+          type: string
+        consumer:
```
Comment (galales) [inline src/main/resources/interface-specification.yml]: Note: the returned model should be exactly the one in the read model, unless we need to hide informations
```suggestion
        consumerId:
```

File: src/main/resources/interface-specification.yml [hunk 5]
```diff
@@ -836,22 +830,58 @@ components:
...
+          type: string
+          format: uuid
+        purposes:
+          type: array
+          items:
+            $ref: '#/components/schemas/ClientStatesChain'
+        relationships:
```
Comment (galales) [inline src/main/resources/interface-specification.yml]:
```suggestion
        relationshipsIds:
```

File: src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiHandlers.scala
Comment (galales) [inline src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiHandlers.scala:1]: ℹ️ remember to delete the old `listClientsResponse`

File: src/test/resources/authz.json
Comment (galales) [inline src/test/resources/authz.json:1]: ℹ️ remember to remove the configuration for `listClients`

File: src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala [hunk 6]
```diff
@@ -708,4 +651,22 @@ final case class ClientApiServiceImpl(
...
+  override def getClients(name: Option[String], relationshipIds: String, offset: Int, limit: Int)(implicit
+    contexts: Seq[(String, String)],
+    toEntityMarshallerClients: ToEntityMarshaller[Clients],
+    toEntityMarshallerProblem: ToEntityMarshaller[Problem]
+  ): Route = authorize(ADMIN_ROLE, SECURITY_ROLE, M2M_ROLE) {
+    val operationLabel =
+      s"Retrieving clients by name $name , relationship $relationshipIds"
+    logger.info(operationLabel)
+
+    val result: Future[Clients] = for {
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala:673]: ⚠️ the [permissions check](https://github.com/pagopa/interop-be-authorization-process/pull/203/files?diff=split&w=1#diff-209aa1b53c41d816589a5ccda66c676d1eb0c523ee48138033a3e7e3d4e27b85L159) has been removed. 
It is required to 
- allow admin users to see all their clients, even if they are not included in the relationships
- if user is not an admin, return only clients that includes the relationship id of the user-entity

**Review by galales — CHANGES_REQUESTED** [review]

File: src/main/scala/it/pagopa/interop/authorizationprocess/common/readmodel/ReadModelQueries.scala [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  private def listClientsFilters(
+    name: Option[String],
+    relationshipIds: List[String],
+    consumerId: String,
+    purposeId: Option[String],
+    kind: Option[String]
+  ): Bson = {
+    val relationshipIdsFilter = mapToVarArgs(relationshipIds.map(Filters.eq("data.relationships", _)))(Filters.or)
+    val nameFilter            = name.map(Filters.regex("data.name", _, "i"))
+    val kindFilter            = kind.map(Filters.regex("data.kind", _, "i"))
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/authorizationprocess/common/readmodel/ReadModelQueries.scala]: No need for regex here
```suggestion
    val kindFilter            = kind.map(k => Filters.eq("data.kind", k.toString))
```

File: src/main/scala/it/pagopa/interop/authorizationprocess/common/readmodel/ReadModelQueries.scala [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        .aggregate[TotalCountResult]("clients", filterPipeline ++ countPipeline, offset = 0, limit = Int.MaxValue)
+    } yield PaginatedResult(results = clients, totalCount = count.headOption.map(_.totalCount).getOrElse(0))
+  }
+
+  private def listClientsFilters(
+    name: Option[String],
+    relationshipIds: List[String],
+    consumerId: String,
+    purposeId: Option[String],
+    kind: Option[String]
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/authorizationprocess/common/readmodel/ReadModelQueries.scala]:
```suggestion
    kind: Option[PersistentClientKind]
```

File: src/main/scala/it/pagopa/interop/authorizationprocess/common/readmodel/ReadModelQueries.scala [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  ): Bson = {
+    val relationshipIdsFilter = mapToVarArgs(relationshipIds.map(Filters.eq("data.relationships", _)))(Filters.or)
+    val nameFilter            = name.map(Filters.regex("data.name", _, "i"))
+    val kindFilter            = kind.map(Filters.regex("data.kind", _, "i"))
+    val consumerFilter        = Filters.eq("data.consumerId", consumerId)
+    val purposeFilter         = purposeId.map(Filters.in("data.purposes", _))
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/authorizationprocess/common/readmodel/ReadModelQueries.scala]: are we sure this is working?
I would expect something like
```suggestion
    val purposeFilter         = purposeId.map(Filters.eq("data.purposes.purpose.purposeId", _))
```

File: src/main/scala/it/pagopa/interop/authorizationprocess/common/readmodel/ReadModelQueries.scala [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+import scala.concurrent.{ExecutionContext, Future}
+
+object ReadModelQueries {
+
+  def listClients(
+    name: Option[String],
+    relationshipIds: List[String],
+    consumerId: String,
+    purposeId: Option[String],
+    kind: Option[String],
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/authorizationprocess/common/readmodel/ReadModelQueries.scala]: let's keep them typed
```suggestion
    relationshipIds: List[UUID],
    consumerId: UUID,
    purposeId: Option[UUID],
    kind: Option[PersistentClientKind],
```

File: src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala [hunk 11]
```diff
@@ -708,4 +650,46 @@ final case class ClientApiServiceImpl(
...
+  ): Route = authorize(ADMIN_ROLE, SECURITY_ROLE, M2M_ROLE) {
+    val operationLabel =
+      s"Retrieving clients by name $name , relationship $relationshipIds"
+    logger.info(operationLabel)
+
+    def getRelationships(userId: UUID, consumerId: UUID): Future[List[String]] = for {
+      relationshipId <- securityOperatorRelationship(consumerId, userId, Seq(SECURITY_ROLE))
+        .map(_.id)
+        .map(_.toString)
+    } yield List(relationshipId)
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala]:
```suggestion
    def getRelationships(userId: UUID, consumerId: UUID): Future[List[UUID]] = 
      securityOperatorRelationship(consumerId, userId, Seq(SECURITY_ROLE)).map(_.id)
```

File: src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala [hunk 12]
```diff
@@ -708,4 +650,46 @@ final case class ClientApiServiceImpl(
...
+
+    val result: Future[Clients] = for {
+      requesterUuid <- getOrganizationIdFutureUUID(contexts)
+      userUuid      <- getUidFutureUUID(contexts)
+      consumerUuid  <- consumerId.toFutureUUID
+      roles         <- getUserRolesFuture(contexts)
+      relationships <-
+        if (roles.contains(SECURITY_ROLE))
+          getRelationships(requesterUuid, userUuid)
+        else Future.successful(parseArrayParameters(relationshipIds))
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala]:
```suggestion
        else parseArrayParameters(relationshipIds).traverse(_.toUuid).toFuture
```

File: src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala [hunk 13]
```diff
@@ -667,24 +667,28 @@ final case class ClientApiServiceImpl(
...
-        .map(_.toString)
-    } yield List(relationshipId)
+    def getRelationship(userId: UUID, consumerId: UUID): Future[UUID] =
+      securityOperatorRelationship(consumerId, userId, Seq(SECURITY_ROLE)).map(_.id)
 
     val result: Future[Clients] = for {
       requesterUuid <- getOrganizationIdFutureUUID(contexts)
       userUuid      <- getUidFutureUUID(contexts)
       consumerUuid  <- consumerId.toFutureUUID
+      purposeUUid   <- purposeId.map(_.toFutureUUID).sequence
```
Comment (TonioGela) [inline src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala]:
```suggestion
      purposeUUid   <- purposeId.traverse(_.toFutureUUID)
```

**Review by TonioGela — APPROVED** [review]

**Review by galales — APPROVED** [review]

**Pull request merged**